File: simple_camera_pid/matlab/experiments/run_turtlebot3_mujoco_pid.m
function out = run_turtlebot3_mujoco_pid( ...
    renderingType, stopTime, targetLateralPosition)
%RUN_TURTLEBOT3_MUJOCO_PID Run the TurtleBot3 PID baseline in MuJoCo.
%
%   out = run_turtlebot3_mujoco_pid("Global", 10) opens the MuJoCo window.
%   out = run_turtlebot3_mujoco_pid("None", 2) runs headless verification.
%   out = run_turtlebot3_mujoco_pid("None", 3, 0.1) tests a nonzero target.

arguments
    renderingType (1,1) string {mustBeMember(renderingType, ...
        ["Global", "Local", "None"])} = "Global"
    stopTime (1,1) double {mustBePositive} = 10
    targetLateralPosition (1,1) double {mustBeFinite} = 0
end

experimentDir = fileparts(mfilename("fullpath"));
modelDir = fileparts(experimentDir);
workspaceDir = fullfile(modelDir, "..", "..", "..");
modelFile = fullfile(modelDir, "turtlebot3_burger_pid_baseline.slx");
xmlFile = fullfile(workspaceDir, "model", "mujoco", "turtlebot3", ...
    "simple_camera_track_turtlebot3_burger_visual_scene.xml");

assert(isfile(modelFile), "MuJoCoPID:MissingModel", ...
    "PID baseline model not found: %s", modelFile);
assert(isfile(xmlFile), "MuJoCoPID:MissingMJCF", ...
    "MuJoCo scene not found: %s", xmlFile);
assert(~isempty(which("mj_initbus")), "MuJoCoPID:MissingBlockset", ...
    "Simulink Blockset for MuJoCo Simulator is not installed.");

[~, model] = fileparts(modelFile);
plant = model + "/MuJoCo_Plant";

in = Simulink.SimulationInput(model);
in = in.setModelParameter("StopTime", num2str(stopTime));
in = in.setVariable("MJ_TargetLateralPosition", ...
    targetLateralPosition, "Workspace", model);
in = in.setBlockParameter(plant, "xmlFileRel", xmlFile);
in = in.setBlockParameter(plant, "xmlFile", xmlFile);
in = in.setBlockParameter(plant, "renderingType", renderingType);

out = sim(in);

odom = out.mujoco_odom_position;
wheel = out.mujoco_wheel_command;
assert(all(isfinite(odom.Data), "all"), ...
    "MuJoCoPID:InvalidOdometry", "MuJoCo odometry contains NaN or Inf.");
assert(all(isfinite(wheel.Data), "all"), ...
    "MuJoCoPID:InvalidWheelCommand", ...
    "Wheel commands contain NaN or Inf.");

odomData = squeeze(odom.Data);
if size(odomData, 1) == 3
    finalPosition = odomData(:, end);
elseif size(odomData, 2) == 3
    finalPosition = odomData(end, :).';
else
    error("MuJoCoPID:InvalidOdometryShape", ...
        "Expected a 3-element position signal, got %s.", ...
        mat2str(size(odom.Data)));
end

fprintf("MuJoCo simulation complete: %.3f s, %d odometry samples.\n", ...
    odom.Time(end), numel(odom.Time));
fprintf("Final position [x y z] = [%.4f %.4f %.4f] m.\n", ...
    finalPosition(1), finalPosition(2), finalPosition(3));
fprintf("Final lateral error = %.4f m.\n", ...
    targetLateralPosition - finalPosition(2));
fprintf("Peak wheel command = %.4f rad/s.\n", ...
    max(abs(wheel.Data), [], "all"));
end

File: simple_camera_pid/simple_camera_pid/mujoco/experiments/run_turtlebot3_mujoco_pid.py
"""Run the TurtleBot3 PID baseline in MuJoCo — headless smoke-test / demo.

Translated from ``run_turtlebot3_mujoco_pid.m``.

Usage
-----
    python3 -m simple_camera_pid.mujoco.experiments.run_turtlebot3_mujoco_pid --stop-time 10
    python3 -m simple_camera_pid.mujoco.experiments.run_turtlebot3_mujoco_pid --stop-time 2 --headless
"""
from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np

from ..sim.turtlebot3_mujoco_env import Turtlebot3LineFollowerEnv


def run_turtlebot3_mujoco_pid(
    repo_root: Path, stop_time: float = 10.0, target_lateral_position: float = 0.0,
) -> Turtlebot3LineFollowerEnv:
    env = Turtlebot3LineFollowerEnv(repo_root, map_key="simple")
    env.reset(options={"lateral_target": target_lateral_position})

    n_steps = round(stop_time / env.controller_cfg.ts_control)
    odom = np.zeros((n_steps, 3))
    wheel = np.zeros((n_steps, 2))
    for k in range(n_steps):
        _, _, terminated, truncated, info = env.step(None)
        log = info["log"]
        odom[k] = log.odom_position
        wheel[k] = log.wheel_command
        if terminated or truncated:
            odom, wheel = odom[: k + 1], wheel[: k + 1]
            break

    if not np.all(np.isfinite(odom)):
        raise RuntimeError("MuJoCo odometry contains NaN or Inf.")
    if not np.all(np.isfinite(wheel)):
        raise RuntimeError("Wheel commands contain NaN or Inf.")

    final_position = odom[-1]
    print(f"MuJoCo simulation complete: {env.data.time:.3f} s, {len(odom)} odometry samples.")
    print(f"Final position [x y z] = [{final_position[0]:.4f} {final_position[1]:.4f} "
          f"{final_position[2]:.4f}] m.")
    print(f"Final lateral error = {target_lateral_position - final_position[1]:.4f} m.")
    print(f"Peak wheel command = {np.max(np.abs(wheel)):.4f} rad/s.")
    return env


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--stop-time", type=float, default=10.0)
    parser.add_argument("--target-lateral", type=float, default=0.0)
    args = parser.parse_args()

    repo_root = Path(__file__).resolve().parents[5]
    run_turtlebot3_mujoco_pid(repo_root, args.stop_time, args.target_lateral)


if __name__ == "__main__":
    main()
